### ISSUE TITLE
Clustering notes

### ISSUE THREAD

**Issue by lefnire** [lefnire/ml-tools#2]
I have clustering implemented via Kmeans and Agglomorative. I've found Agglomorative substantially better. To use Kmeans effectively, I needed to dimensionality-reduce first (see `ae.py`). Dim-reducing via T-SNE, PCA, Umap were very ineffective for me. I'm considering this solved for now, using Agglomorative as my default, but I'll dump my old notes here in case this wants to be re-discussed. Old notes:

Clustering
- wait for https://github.com/jgraving/vaesne
- joint loss
  - Don't do argmax topic, do full prob output - softmax - cat_xentropy. 
    - Combine cosine score with Jensen (in network? Later in cdist?)
  - joint umap(gmm)?
    - use umap.euclidean (preserve score), umap(gmm) for themes. remove lda
  - mse x (autoencoder)
    - switch to roberta_large
  - cdist rand_other_x
    - attached to output, to maintain cdist in encoding
  - softmax label
    - from LDA, coherence_score_ (n_clusters=n_topics)
    - fix up spacy cleanup
      "https://towardsdatascience.com/setting-up-text-preprocessing-pipeline-using-scikit-learn-and-spacy-e09b9b76758f"
      - Don't remove punct, do in lemma 
      - Unlemma in link
- cosine clustering
  - kmedoid can't fit in RAM. Trying random sampling, too much beef
  - also tried hierarchical, dbscan, hdbscan, etc. with cosine
- train AE on nli-stbs datase w triplet loss, to ensure encoder accurate / keeps structure
  - Joint loss AE output w lda topics topic-adsign softmax (n_topics is n_clusters)
- umap seems to give good distortion/bic, but terrible book results
- gmm(umap) is terrible! stick to AE+kmeans
  - [COMPLETE] if gmm(umap) is much better (subjective, bic); then save its assignments as y_true, pipe into encoder aux_output to train dense(1)
